pallets/drop3/src/lib.rs

```diff
@@ -0,0 +1,514 @@
+			// pools and the balance is reserved/unreserved here and there.
+			//
+			// If it somehow fails, it implies somewhere the code logic is wrong or the the reserved
+			// balance was (unexpectedly) unreserved
+			//
+			// we don't care so much about the comparison between reserved balance and pool.remain,
+			// the worst case is pool.remain is higher than reserved balance when unreserving, which
+			// is not so bad as up to 'reserved balance' will be unreserved anyway.
+			ensure!(
+				<pallet_balances::Pallet<T> as ReservableCurrency<_>>::reserved_balance(
+					&pool.owner
+				) >= amount,
+				Error::<T>::InsufficientReservedBalance
+			);
+			// do the transfer from the reserved balance
+			// we shall make sure the correct amount is included in event even when
+			// unmoved is non-zero (which should not happen)
+			let unmoved =
+				<pallet_balances::Pallet<T> as ReservableCurrency<_>>::repatriate_reserved(
+					&pool.owner,
+					&to,
+					amount,
+					BalanceStatus::Free,
+				)?;
+			let actual_moved = amount - unmoved;
+			Self::deposit_event(Event::RewardSent { to, amount: actual_moved });
+			RewardPools::<T>::mutate(id, |pool| {
+				pool.remain = pool.remain.saturating_sub(actual_moved);
+			});
+			ensure!(unmoved == Zero::zero(), Error::<T>::UnexpectedUnMovedAmount);
+			Ok(().into())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// close the reward pool, remove it from the map
+		/// and unreserve all the remaining token to the pool owner
+		fn unreserve_and_close_reward_pool(id: T::PoolId) {
+			let pool = RewardPools::<T>::take(id);
+			// we don't care if reserved balance is less than pool.remain
+			let _ = <pallet_balances::Pallet<T> as ReservableCurrency<_>>::unreserve(
+				&pool.owner,
+				pool.remain,
+			);
+			RewardPoolOwners::<T>::remove(pool.id);
+
+			Self::deposit_event(Event::RewardPoolRemoved {
+				id: pool.id,
+				name: pool.name.into_inner(),
+				owner: pool.owner,
+			});
+		}
+
+		pub fn get_sorted_pool_ids() -> Vec<T::PoolId> {
+			let mut ids = RewardPools::<T>::iter_keys().collect::<Vec<T::PoolId>>();
+			ids.sort();
+			ids
+		}
+
+		#[cfg(test)]
+		// propose a default reward pool, but with given id
+		// mainly used to test get_next_pool_id()
+		pub fn propose_default_reward_pool(id: T::PoolId, should_change_current_max: bool) {
+			RewardPools::<T>::insert(id, RewardPool::default());
+			RewardPoolOwners::<T>::insert(id, T::AccountId::default());
+			if should_change_current_max {
+				CurrentMaxPoolId::<T>::put(id);
+			}
+		}
+
+		fn get_next_pool_id() -> Result<T::PoolId, pallet::Error<T>> {
+			// if CurrentMaxPoolId hasn't reached max, increment and return it
+			if Self::current_max_pool_id() < T::PoolId::max_value() {
+				return CurrentMaxPoolId::<T>::mutate(|id| {
+					*id += 1u64.into();
+					Ok(*id)
+				})
+			}
+
+			// otherwise find the vacant id from the beginning
+			let sorted_ids = Self::get_sorted_pool_ids();
```

Comment:
u64 is big enough, if necessary to find a used id after reach max value.

Reply:
> u64 is big enough, if necessary to find a used id after reach max value.

Had this thought before, just want to play it safe and cover all cases.
It's unlikely to happen (but still possible as anyone can call `propose_reward_pool` with a minimum transaction fee and reserved balance), but once the id is exhausted, it needs a runtime upgrade just for changing this.